tests/src/test/java/org/eclipse/hono/tests/registry/CorsIT.java

```diff
@@ -0,0 +1,224 @@
+        doPreflightCheck(uri, HttpMethod.GET, ctx);
+    }
+
+    /**
+     * Verifies that the device registry returns matching CORS headers in response to a
+     * CORS preflight request for deleting device data.
+     *
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    public void testCorsPreflightRequestForDeletingDevice(final TestContext ctx) {
+
+        final String uri = String.format(TEMPLATE_URI_REGISTRATION_INSTANCE, DEFAULT_TENANT, "my-device");
+        doPreflightCheck(uri, HttpMethod.DELETE, ctx);
+    }
+
+    /**
+     * Verifies that the HTTP adapter returns matching CORS headers in response to a
+     * CORS preflight request for updating credentials data.
+     *
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    public void testCorsPreflightRequestForUpdatingCredentials(final TestContext ctx) {
+
+        final String uri = String.format(TEMPLATE_URI_CREDENTIALS_BY_DEVICE, DEFAULT_TENANT, "my-device");
+        doPreflightCheck(uri, HttpMethod.PUT, ctx);
+    }
+
+    /**
+     * Verifies that the device registry returns matching CORS headers in response to a
+     * CORS preflight request for getting credentials data.
+     *
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    public void testCorsPreflightRequestForGettingCredentials(final TestContext ctx) {
+
+        final String uri = String.format(TEMPLATE_URI_CREDENTIALS_BY_DEVICE, DEFAULT_TENANT, "my-device");
+        doPreflightCheck(uri, HttpMethod.GET, ctx);
+    }
+
+    private void doPreflightCheck(final String uri, final HttpMethod method, final TestContext ctx) {
+
+        httpClient.options(
+                uri,
+                MultiMap.caseInsensitiveMultiMap()
+                    .add(HttpHeaders.ORIGIN, CrudHttpClient.ORIGIN_URI)
+                    .add(HttpHeaders.ACCESS_CONTROL_REQUEST_METHOD, method.name()),
+                status -> status == HttpURLConnection.HTTP_OK)
+        .setHandler(ctx.asyncAssertSuccess(headers -> {
+            assertAccessControlHeaders(ctx, headers, method);
+        }));
+    }
+
+
+    private static void assertAccessControlHeaders(final TestContext ctx, final MultiMap headers, final HttpMethod expectedAllowedMethod) {
+
+        ctx.assertTrue(headers.get(HttpHeaders.ACCESS_CONTROL_ALLOW_METHODS).contains(expectedAllowedMethod.name()));
+        ctx.assertEquals("*", headers.get(HttpHeaders.ACCESS_CONTROL_ALLOW_ORIGIN));
+        ctx.assertTrue(headers.get(HttpHeaders.ACCESS_CONTROL_ALLOW_HEADERS).contains(HttpHeaders.AUTHORIZATION));
+        ctx.assertTrue(headers.get(HttpHeaders.ACCESS_CONTROL_ALLOW_HEADERS).contains(HttpHeaders.CONTENT_TYPE));
+        ctx.assertTrue(headers.get(HttpHeaders.ACCESS_CONTROL_ALLOW_HEADERS).contains(HttpHeaders.IF_MATCH));
+        ctx.assertTrue(headers.get(HttpHeaders.ACCESS_CONTROL_ALLOW_HEADERS).contains(HttpHeaders.ETAG));
```

Comment:
this is not supposed to show up in the allowed headers but in the exposed headers instead

Reply:
Reading the spec, it says that server returns allow headers as a part of the preflight check. So this seems OK to me. Exposed headers are (as far as I can see) returned as the part of the regular calls. So this looks OK to me. The question is whether we should add these headers to the list of exposed headers. Looking at the HTTP adapter, it does the same thing as here.

Reply:
I agree, in the preflight check, the server should return the list of headers in the *allowed headers* that the client may include in its request, right? Since the *ETag* header is only included in responses from the server, it shouldn't be included here FMPOV. The pre-flight request shouldn't include it in its *allowed headers* either.
This, of course, needs to be changed in conjunction with canging the definition of the CorsHandler in the AbstractHttpEndpoint as noted in an earlier comment: *ETag* should be added as an *exposed header*.

Reply:
OK, now that made sense :) Fixed!